bodo/pandas/physical/limit.h

```diff
@@ -20,22 +21,43 @@ class PhysicalLimit : public PhysicalSourceSink {
     /**
      * @brief Do limit.
      *
-     * @return std::pair<int64_t, OperatorResult> Bodo C++ table pointer cast to
-     * int64 (to pass to Cython easily), state of operator after processing the
-     * batch
+     * @return std::pair<std::shared_ptr<table_info>, OperatorResult>
+     * The output table from the current operation and whether there is more
+     * output.
      */
     std::pair<std::shared_ptr<table_info>, OperatorResult> ProcessBatch(
         std::shared_ptr<table_info> input_batch) override {
-        uint64_t select_this_time = std::min(n, input_batch->nrows());
-        n -= select_this_time;
+        int n_pes = 0;
+        int myrank = 0;
+        MPI_Comm_size(MPI_COMM_WORLD, &n_pes);
+        MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
 
-        std::vector<int64_t> rowInds(select_this_time);
-        for (uint64_t i = 0; i < select_this_time; ++i) {
+        // Gather how many rows are on each rank
+        std::vector<uint64_t> row_counts(n_pes);
+        uint64_t cur_rows = input_batch->nrows();
+        CHECK_MPI(MPI_Allgather(&cur_rows, 1, MPI_UNSIGNED_LONG_LONG,
```

Comment:
Using blocking collectives in the middle of pipeline loop is not safe since ranks may have different number of iterations (e.g. some rank has more data). I think using async collectives is the proper approach but to limit the scope right now, each rank can generate `n` rows at most and move this code to the finalize function of the operator.

Reply:
We should talk.